src/library.js

```diff
@@ -3266,6 +3268,42 @@ LibraryManager.library = {
     return ASM_CONSTS[code].apply(null, args);
   },
   emscripten_asm_const_double: 'emscripten_asm_const_int',
+  emscripten_asm_const_js_val__sig: 'iiii',
+  emscripten_asm_const_js_val__deps: ['emscripten_asm_const_int', '$JsValStore'],
+  emscripten_asm_const_js_val: function(code, sigPtr, argbuf) {
+    return JsValStore.add(_emscripten_asm_const_int(code, sigPtr, argbuf));
+  },
+
+  $JsValStore: {
+    values: [undefined],
+    free_ids: [],
+    next_id: 1,
+
+    add: function(js_val) {
+      const id = JsValStore.free_ids.length ? JsValStore.free_ids.pop() : JsValStore.next_id++;
```

Comment:
This code has the effect of reusing the same ID out immediately after it has been freed, which can mask errors. I.e. if one allocates a buffer A, then frees A, then allocates buffer B - then if the code erroneously still accesses buffer A ptr, it will actually do modifications to buffer B, since A and B will have the same ID.

Reply:
Ok - I've implemented the same wraparound handling here and removed the free list.